src/toplevel/virtualprocess.jl

```diff
@@ -812,22 +843,24 @@ end
 
 function macroexpand_with_err_handling(state::InterpretationState, x::Expr)
     # `scrub_offset = 2`: `macroexpand` -> kwfunc (`macroexpand`)
-    with_err_handling(general_err_handler, state; scrub_offset=2) do
-        # XXX we want to non-recursive, sequential partial macro expansion here, which allows
-        # us to collect more fine-grained error reports within macro expansions
+    with_err_handling(macro_expansion_err_handler, state; scrub_offset=2) do
+        # XXX we want to non-recursive, sequential partial macro expansion here,
+        # which allows us to collect more fine-grained error reports within macro expansions
         # but it can lead to invalid macro hygiene escaping because of https://github.com/JuliaLang/julia/issues/20241
-        macroexpand(state.context, x; recursive = true #= but want to use `false` here =#)
+        macroexpand(state.context, x; recursive = true)
     end
 end
 
 function lower_with_err_handling(interp::ConcreteInterpreter, ::JS.SyntaxNode, x::Expr)
     # `scrub_offset = 1`: `lower`
     state = InterpretationState(interp)
-    with_err_handling(general_err_handler, state; scrub_offset=1) do
-        lwr = lower(state.context, x)
+
+    x_expanded = @something macroexpand_with_err_handling(state, x) return nothing
+
+    with_err_handling(lowering_err_handler, state; scrub_offset=1) do
+        lwr = lower(state.context, x_expanded)
         if isexpr(lwr, :error)
-            # here we should capture syntax errors found during lowering
-            msg = first(lwr.args)::String
+            msg = first(lwr.args)
             add_toplevel_error_report!(state, LoweringErrorReport(msg, state.filename, state.curline))
```

Comment:
@aviatesk it seems this line (branch?) should have updated as well. AFAICT it causes https://github.com/aviatesk/JET.jl/issues/748 (which I just happened to run into as well).